### PR TITLE
Sub: Lua scripts can query joystick button state

### PR DESCRIPTION
**Pull request by clydemcqueen**
This PR provides a way to use a joystick (used by Sub) to provide input to Lua scripts. You can associate a counter with a joystick button, and this counter can be queried and cleared from inside Lua scripts. This provides an easy way to trigger various kinds of script behaviors using a joystick. The example script provides documentation. Feedback welcome.

### CODE REVIEW

**Comment by rmackay9** [comment]
Thanks, it would be good to fixup the commits to start with the subsystem.  E.g "Sub:", "AP_JSButton" and "AP_Scripting:".  This makes it easier for us to quickly see which area of the system is affected by the change and helps with backporting.

**Comment by rmackay9** [comment]
As a side note, I can't help but wonder, if this could be avoided by adding the RC_Channel library to Sub.

**Comment by clydemcqueen** [comment]
Thanks @rmackay9, I can certainly amend && re-arrange the commits:
* AP_JSButton: reserve 4 button functions for script access
* AP_Scripting: add sub bindings for joystick buttons
* Sub: provide script access to joystick buttons

**Comment by clydemcqueen** [comment]
I'm not sure why test.BalanceBot is failing, it succeeds locally:
~~~
./Tools/autotest/autotest.py build.Rover test.BalanceBot
...
>>> PASSED STEP: test.BalanceBot at Tue Aug 15 11:29:45 2023
~~~

**Review by IamPete1 — COMMENTED** [review]
So far the pattern for getting to vehicle stuff has been to go via AP Vehicle. This would be the first case of adding a vehicle singleton directly to scripting. Quadplane has one, but that only half counts. 

Personally I like this pattern better, but it is different to what we have done in the past. 

If we were to add support for JSButtons to other vehicles it would be better to add the singleton to `AP_JSButton` and handle scripting directly.

File: ArduSub/script_button.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+//
+class ScriptButton {
+public:
+    ScriptButton(): pressed(false), count(0) {}
+
+    void press();
+
+    void release();
+
+    [[nodiscard]] bool is_pressed() const;
```
Comment (IamPete1) [inline ArduSub/script_button.h:21]:
```suggestion
    bool is_pressed() const WARN_IF_UNUSED;
```
We have a macro for this that we use else where.

Comment (clydemcqueen) [reply ArduSub/script_button.h:21]: Thanks, I will update.

File: ArduSub/Sub.cpp [hunk 2]
```diff
@@ -43,7 +43,13 @@ Sub::Sub()
 #if CONFIG_HAL_BOARD != HAL_BOARD_SITL
     failsafe.pilot_input = true;
 #endif
+    if (_singleton != nullptr) {
+        AP_HAL::panic("Can only be one Sub");
+    }
```
Comment (rmackay9) [inline ArduSub/Sub.cpp:48]: This should be protected by a definition to be sure this panic only occurs SITL.

**Review by tridge — APPROVED** [review]

**Review by Williangalvani — APPROVED** [review]

**Pull request merged**